### PR TITLE
Added processment of data models to the blog

### PR DESCRIPTION
**Pull request by ghost**
Create models

    Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
    Set up associations between models.

Use models to insert data

    Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
    Open rails console for your project.
    Create at least one user by running the following code:

       first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
       second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')

    Create at least 4 posts written by one of the users you created by running the following code:

       first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')

    Create at least 6 posts comments for one of the posts you created by running the following code:

       Comment.create(post: first_post, author: second_user, text: 'Hi Tom!' )

    Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.

Create custom methods

Your models should include some additional methods.

    Users
        A method that returns the 3 most recent posts for a given user.
    Posts
        A method that updates the posts counter for a user.
        A method which returns the 5 most recent comments for a given post.
    Comments
        A method that updates the comments counter for a post.
    Likes
        A method that updates the likes counter for a post.
    Go to rails c and check if your methods are working.

Additionally, you will need to make current user data available to all controllers.

    Create a method called current_user in your ApplicationController.
    A current_user method should return the first user from a database.
    Don't worry about proper authentication at this time. You will add it in the upcoming projects.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -3,4 +3,5 @@ class User < ApplicationRecord
   has_many :posts, dependent: :destroy
   has_many :likes, dependent: :nullify
```
Comment (arslanbisharat) [inline app/models/user.rb]: - [ ] The associations between user models are set up correctly but according to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png) you should have foreign keys 'author id' for each association.

File: app/models/post.rb [hunk 2]
```diff
@@ -2,4 +2,6 @@ class Post < ApplicationRecord
   belongs_to :user
```
Comment (arslanbisharat) [inline app/models/post.rb]: - [ ] According to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png) you should have post.rb belongs to the author and it should have classname 'User'.

File: app/models/comment.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
 class Comment < ApplicationRecord
   belongs_to :user
   belongs_to :post
```
Comment (arslanbisharat) [inline app/models/comment.rb]: - [ ] According to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png) you should have comment.rb belongs to the author and it should have classname 'User'.

File: app/models/like.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
 class Like < ApplicationRecord
   belongs_to :user
   belongs_to :post
```
Comment (arslanbisharat) [inline app/models/like.rb]: - [ ] According to the given [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png) you should have like.rb belongs to the author and it should have classname 'User'.

File: app/models/post.rb [hunk 5]
```diff
@@ -2,4 +2,6 @@ class Post < ApplicationRecord
...
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+    scope :update_posts_counter, ->(user, count) { user.update(posts_counter: count) }
+  scope :last_five_comments, ->(post) { Comment.where(post:).last(5) }
```
Comment (arslanbisharat) [inline app/models/post.rb]: - [ ] According to the guidelines you should have a method that updates the posts counter for a user. You can check out the guide [Scope vs Class Methods](https://www.square63.com/scope-vs-class-methods/)

```ruby
def update_post_counter
  author.increment!(:posts_counter)
end
```

- [ ] [OPTIONAL] You can use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to database, then make the `update_post_counter` method private

```ruby
after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/user.rb [hunk 6]
```diff
@@ -3,4 +3,5 @@ class User < ApplicationRecord
...
   validates :name, presence: true
+    scope :last_three_posts, ->(user) { Post.where(author: user).last(3) }
```
Comment (arslanbisharat) [inline app/models/user.rb]: - [ ] According to the guidelines you should have a method that returns 3 most recent posts. You can checkout the guide [Scope vs Class Methods](https://www.square63.com/scope-vs-class-methods/)

```ruby
def recent_post
  posts.order('created_at Desc').limit(3)
end
```

File: app/models/comment.rb [hunk 7]
```diff
@@ -1,4 +1,5 @@
...
+    scope :update_comments_counter, ->(post, count) { post.update(comments_counter: count) }
```
Comment (arslanbisharat) [inline app/models/comment.rb]: - [ ] According to the guidelines you should have a method that returns the 5 most recent comments for a given post. You can check out the guide [Scope vs Class Methods](https://www.square63.com/scope-vs-class-methods/)

```ruby
def recent_comments
  comments.order('created_at Desc').limit(5)
end
```

File: app/models/like.rb [hunk 8]
```diff
@@ -1,4 +1,5 @@
...
+  scope :update_likes_counter, ->(post, count) { post.update(likes_counter: count) }
```
Comment (arslanbisharat) [inline app/models/like.rb]: - [ ] According to the guidelines you should have a method that updates the likes counter for a post. You can check out the guide [Scope vs Class Methods](https://www.square63.com/scope-vs-class-methods/)

```ruby
def update_likes_counter
  post.increment!(:likes_counter)
end
```

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
Hi @iKuartz, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

- The Pull Request has a proper title and description. ✔️
- You have added a descriptive Readme file. Good Job. ✔️
- There is no linter error is present. ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [ ] Please make sure you right descriptive comments. Commenting involves placing Human Readable Descriptions inside of computer programs detailing what the Code is doing. Proper use of commenting can make code maintenance much easier, as well as helping make finding bugs faster. Further, commenting is very important when writing functions that other people will use. 🟢 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by ghost** [comment]
@arslanbisharat Thank you very much for your careful observation and helpful comments. Your review is correct, I really must refactor this code, and the way you put it and referenced material I really have to take another look at actually helped me. I am re-doing the code at the moment. Thank you.

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:

### Hi Ivan, 
## Congratulations 🎉 Thank you for implementing all the requirements 🙏
### Please go ahead and merge the branch!

## Happy Coding 👏👏👏
#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**